p2p/protocol/identify/id.go

```diff
@@ -197,11 +197,29 @@ func (ids *IDService) IdentifyConn(c network.Conn) {
 		delete(ids.currid, c)
 		ids.currmu.Unlock()
 
-		// emit the appropriate event.
-		if p := c.RemotePeer(); err == nil {
-			ids.emitters.evtPeerIdentificationCompleted.Emit(event.EvtPeerIdentificationCompleted{Peer: p})
+		// emit the appropriate events.
+		rp := c.RemotePeer()
+		if err == nil {
+			// emit Identify Completed
+			ids.emitters.evtPeerIdentificationCompleted.Emit(event.EvtPeerIdentificationCompleted{Peer: rp})
+
+			// emit Peer Protocols Updated
+			peerProtocols, err := ids.Host.Peerstore().GetProtocols(rp)
+			if err != nil {
+				log.Errorf("failed to fetch protocols from the peerstore for peer=%s, err=%s", rp.String(), err)
+				return
+			}
+
+			// TODO This can race with the ID Delta if the peer immediately changes it's protocols after connecting
+			// The long term solution is to fix https://github.com/libp2p/go-libp2p/issues/823
+			// But, in the long term, we are all dead.
+			ids.emitters.evtPeerProtocolsUpdated.Emit(event.EvtPeerProtocolsUpdated{
+				Peer:  c.RemotePeer(),
+				Added: protocol.ConvertFromStrings(peerProtocols),
```

Comment:
Hm. I'm not sure if this is accurate, especially if we have multiple concurrent connections and/or if we're doing an identify push. In that case, we'll be removing/adding protocols.

Really, I wonder if tracking the added/removed protocols is more trouble than it's worth. Accurately tracking this with no race conditions seems horribly difficult unless we synchronously emit updates from the peerstore itself.

Thoughts @raulk? My vote would be to simply remove Added/Removed and expect subscribers to just check the peerstore for the latest values.

Reply:
+1 for subscribers checking the peerstore.

Reply:
fwiw, in JS we emit all protocol and address change events from the peerstore itself to keep a single source of truth for peer records. We also just emit the full, current list of protocols/multiaddrs of the peer to keep the logic simple.

Reply:
@jacobheun 's comment is also a +1 for getting in https://github.com/libp2p/go-libp2p/issues/802.

Reply:
Single source of truth is what we're shooting for. Unfortunately, the peerstore was poorly architected/built/integrated/used from the outset in go-libp2p, and we have to deal with that heritage until v2.

Reply:
> Really, I wonder if tracking the added/removed protocols is more trouble than it's worth.

It is, when this event is used improperly. See https://github.com/libp2p/go-libp2p/issues/860#issuecomment-607203490.

Reply:
To summarise:

- The current source of truth for "peer protocols updated" event is **identify delta**.
- Via identify delta, the peer reports _exactly_ which protocols were added and removed. Those fields are mapped onto the event.
- Whether we emit it from identify delta, or in a future v2 peerstore, the behaviour will be the same.
- This problem arises because this PR emits the event from an improper place.
- This event is an application of CDC (Change Data Capture): https://en.wikipedia.org/wiki/Change_data_capture.